develop/codecs.md

```diff
@@ -0,0 +1,440 @@
+    Codec.INT.fieldOf("beans_amount").forGetter(CoolBeansClass::getBeansAmount),
+    Registries.ITEM.getCodec().fieldOf("bean_type").forGetter(CoolBeansClass::getBeanType),
+    BlockPos.CODEC.listOf().fieldOf("bean_positions").forGetter(CoolBeansClass::getBeanPositions)
+    // Up to 16 fields can be declared here
+).apply(instance, CoolBeansClass::new));
+```
+
+Each line in the group specifies a codec, a field name, and a getter method. The `Codec#fieldOf` call is used to convert
+the codec into a [map codec](#mapcodec-not-to-be-confused-with-codecltmapgt), and the `forGetter` call specifies the getter 
+method used to retrieve the value of the field from an instance of the class. Meanwhile, the `apply` call specifies the 
+constructor used to create new instances. Note that the order of the fields in the group should be the same as the order 
+of the arguments in the constructor.
+
+You can also use `Codec#optionalFieldOf` in this context to make a field optional, as explained in
+the [Optional Fields](#optional-fields) section.
+
+### MapCodec, not to be confused with Codec&lt;Map&gt;
+
+Calling `Codec#fieldOf` will convert a `Codec<T>` into a `MapCodec<T>`, which is a variant, but not direct
+implementation of `Codec<T>`. `MapCodec`s, as their name suggests are guaranteed to serialize into a 
+key to value map, or its equivalent in the `DynamicOps` used. Some functions may require one over a regular codec.
+
+This particular way of creating a `MapCodec` essentially boxes the value of the source codec 
+inside a map, with the given field name as the key. For example, a `Codec<BlockPos>`
+when serialized into json would look like this:
+
+```json
+[1, 2, 3]
+```
+
+But when converted into a `MapCodec<BlockPos>` using `BlockPos.CODEC.fieldOf("pos")`, it would look like this:
+
+```json
+{
+  "pos": [1, 2, 3]
+}
+```
+
+While the most common use for map codecs is to be merged with other map codecs to construct a codec for a full class worth of
+fields, as explained in the [Merging Codecs for Record-like Classes](#merging-codecs-for-record-like-classes) section
+above, they can also be turned back into regular codecs using `MapCodec#codec`, which will retain the same behavior of
+boxing their input value.
+
+#### Optional Fields
+
+`Codec#optionalFieldOf` can be used to create an optional map codec. This will, when the specified field is not present
+in the container during deserialization, either be deserialized as an empty `Optional` or a specified default value.
+
+```java
+// Without a default value
+MapCodec<Optional<BlockPos>> optionalCodec = BlockPos.CODEC.optionalFieldOf("pos");
+
+// With a default value
+MapCodec<BlockPos> optionalCodec = BlockPos.CODEC.optionalFieldOf("pos", BlockPos.ORIGIN);
+```
+
+Do note that optional fields will silently ignore any errors that may occur during deserialization. This means that if
+the field is present, but the value is invalid, the field will always be deserialized as the default value.
```

Comment:
Minecraft (not DFU!) provides a method that creates an optional codec that errors if it failed to deserialize. Might worth mentioning it here.

[`Codecs#createStrictOptionalFieldCodec`](https://maven.fabricmc.net/docs/yarn-1.20.4+build.3/net/minecraft/util/dynamic/Codecs.html#createStrictOptionalFieldCodec(com.mojang.serialization.Codec,java.lang.String))